kaldb/src/main/java/com/slack/kaldb/chunkManager/ChunkCleanerService.java

```diff
@@ -1,54 +1,89 @@
  */
-public class ChunkCleanerService<T> extends AbstractScheduledService {
+public class ChunkCleanerService<T> {
 
   private static final Logger LOG = LoggerFactory.getLogger(ChunkCleanerService.class);
 
+  private final ChunkManager<T> chunkManager;
+  private final int limit;
+
   private final Duration staleDelayDuration;
-  private final IndexingChunkManager chunkManager;
 
-  public ChunkCleanerService(IndexingChunkManager chunkManager, Duration staleDelayDuration) {
+  public ChunkCleanerService(ChunkManager<T> chunkManager, int limit, Duration staleDelayDuration) {
     this.chunkManager = chunkManager;
+    this.limit = limit;
     this.staleDelayDuration = staleDelayDuration;
   }
 
-  @VisibleForTesting
-  public void runAt(Instant instant) {
-    LOG.info("Starting ChunkCleanerTask.");
-    deleteStaleData(instant);
-    LOG.info("Finished ChunkCleanerTask.");
+  public int deleteStaleData(Instant startInstant) {
+    int total = 0;
+    final long staleCutOffMs =
+        startInstant.minusSeconds(staleDelayDuration.toSeconds()).toEpochMilli();
+    total += deleteStaleChunksPastCutOff(staleCutOffMs);
+    total += deleteChunksOverLimit(this.limit);
+    return total;
   }
 
-  @VisibleForTesting
-  public int deleteStaleData(Instant startInstant) {
+  public int deleteStaleData() {
+    int total = 0;
+    Instant startInstant = Instant.now();
     final long staleCutOffMs =
         startInstant.minusSeconds(staleDelayDuration.toSeconds()).toEpochMilli();
-    return deleteStaleChunksPastCutOff(staleCutOffMs);
+    total += deleteStaleChunksPastCutOff(staleCutOffMs);
+    total += deleteChunksOverLimit(this.limit);
+    return total;
+  }
+
+  private int deleteChunksOverLimit(int limit) {
+    if (limit < 0) {
+      throw new IllegalArgumentException("limit can't be negative");
+    }
+
+    final List<Chunk<T>> unsortedChunks = chunkManager.getChunkList();
+
+    if (unsortedChunks.size() <= limit) {
+      LOG.info("Unsorted chunks less than or equal to limit. Doing nothing.");
+      return 0;
+    }
+
+    // Sorts the list in ascending order (i.e. oldest to newest) and only gets chunks that we've
+    // taken a snapshot of
+    final List<Chunk<T>> sortedChunks =
+        unsortedChunks.stream()
+            .sorted(Comparator.comparingLong(chunk -> chunk.info().getChunkCreationTimeEpochMs()))
+            .filter(chunk -> chunk.info().getChunkSnapshotTimeEpochMs() > 0)
+            .toList();
+
+    final int totalChunksToDelete = sortedChunks.size() - limit;
+
+    final List<Chunk<T>> chunksToDelete = sortedChunks.subList(0, totalChunksToDelete);
+
+    LOG.info("Number of chunks past limit of {} is {}", limit, chunksToDelete.size());
+    chunkManager.removeStaleChunks(chunksToDelete);
+    return chunksToDelete.size();
   }
 
-  @VisibleForTesting
-  public int deleteStaleChunksPastCutOff(long staleDataCutOffMs) {
+  private int deleteStaleChunksPastCutOff(long staleDataCutOffMs) {
```

Comment:
In general taking `Instant` instead of a `long` when referring to time is a much safer design - as you don't have to worry about someone in the future passing an incorrect time unit.